### ISSUE TITLE
[Implementation] Indexes' laziness

### ISSUE THREAD

**Issue by Yomguithereal** [medialab/graphlib#16]
Some indexes, notably the neighbors one, can easily made lazy to save some memory and computation time for use cases that don't use them.

That is to say that, they won't be computed before the user calls `neighbors`, for instance.

**Comment by boogheta** [comment]
For performance reasons, one might prefer that it always happens at first instantiation, so having an option in the main constructor would be best

**Comment by Yomguithereal** [comment]
Yup. The idea is to provide an option like `true` `false` and `lazy`.

**Comment by jacomyma** [comment]
The true/false/lazy switch seems a very nice idea in principle. Note that the benefit of computing it just for the necessary nodes / edges is counterbalanced by the fact that it is often more efficient to compute it either "online" or for the whole graph.
- (in-/out-)degree, (in-/out-)edges index, neighbors index, pageRank (...) can be computed "online": once at initialization and updated at modification. It is efficient to compute them in background.
- betweenness centrality, eccentricity, closeness centrality, shortest paths, diameter (...) are not "online" but have to be computed globally for the whole graph to be efficient. So either you don't use them or you take the (computational) time to obtain them.
Note: I would have to recheck exactly those properties for these indexes and metrics.

**Comment by boogheta** [comment]
I guess the idea for the not online ones would be that lazy mode only computes them all when trying to access one of them for the first time.

Not sure of the interests of the `false` option though : the only usecase is when you don't want to ue any of them, and you can never really know that beforehand, and in thi case `lazy` fits your need as it won't take any more computing time as long as the user does not use any

**Comment by Yomguithereal** [comment]
> I guess the idea for the not online ones would be that lazy mode only computes them all when trying to access one of them for the first time.

Yes, for some indexes, computing the index for one node is as complex as computing it for every nodes. The difference here is rather about memory.

I guess `false` would mean "do not compute beforehand" and therefore would mean the aforementioned kind of lazy.

Then "lazy" should probably apply to compute only for the relevant parts of the graph I guess...

On a side note, the graph should also come with methods triggering explicitly the computation of some index if it has not be computed yet. Something in the idea of `graph.computeTheDamnIndex()`.
